Fix west rtt for shell interaction

MarkoSagadin opened the pull request:
## Description

This PR fixes shell RTT interactions via west rtt command. Previously, it was only suitable for presenting logs captured via the RTT, meaning that arrow and navigation keys weren't correctly interpreted. Separate Unix and Windows implementations were needed to resolve this. More details can be found in the commit message.


**Note**: The commit in this PR was moved from the #96822 as per request from this [comment](https://github.com/zephyrproject-rtos/zephyr/pull/96822#issuecomment-3572412537).

MarkoSagadin committed:
scripts: west_commands: runners: fix west rtt shell interaction

When the west rtt command is executed, two things happen:
- A GDB server is run.
- A telnet client (via sockets) is opened to connect to that server.

Previously, that telnet client was operating in the line mode in both
directions. That mode is perfectly fine for just presenting the captured
content (RTT data) to the stdout; however, it is not suitable for
interacting with the device via the shell subsystem. None of the
control commands, such as tab and arrow keys, were correctly passed down
to the device.

To solve that the telnet client had to be switched into the character
mode. For that two separate implementations were needed, one for the
Unix and one for Windows. That was solved with two different context
managers.

They both wrap the common logic: two threads, where one reads input
from the keyboard and sends it to the telnet client, and the other on
reads the data from the telnet client and prints it to the screen.

Signed-off-by: Marko Sagadin <marko.sagadin42@gmail.com>

MarkoSagadin commented:
Remove stale label, I still care about getting this PR merged.

marc-hb reviewed (commented):
> Since we don't have tty package on the Windows, ...

Maybe I'm missing something but I think this has been available for a long time on Windows, it's just not on by default:

https://stackoverflow.com/questions/36760127/how-to-use-the-new-support-for-ansi-escape-sequences-in-the-windows-10-console

From https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences

```
Sequence | Code | Description | Behavior
-- | -- | -- | --
ESC [ <n> A | CUU | Cursor Up | Cursor up by <n>
ESC [ <n> B | CUD | Cursor Down | Cursor down by <n>
...
```

etc.

MarkoSagadin commented:
Hi @marc-hb, I have converted the windows-specific function to use the `ENABLE_VIRTUAL_TERMINAL_INPUT` and `ENABLE_VIRTUAL_TERMINAL_PROCESSING` from the `windll` instead of the `KEY_MAP` dict.

Please take another look, this should be ok now.

marc-hb reviewed (commented):
Would Linux and macOS lose anything if they were using the same two threads instead of `select()` ? Without the kernel32.setconsolemode() and friends of course. This would considerably reduce the divergence and OS-dependent code, allowing anyone to troubleshoot and fix most issues for any OS.

MarkoSagadin commented on scripts/west_commands/runners/core.py:
```diff
@@ -1028,20 +1029,165 @@ def run_telnet_client(
-                elif key.fileobj == sock:
+
+
+        # We can't type check functions from tty and termios on Windows operating
```
Shouldn't be here, I have removed it.

MarkoSagadin commented:
Hi @marc-hb, I addressed both of your suggestions. The platform specific code is now solved with context managers, the rest of the code uses the the same read and write loops for the all platforms. Threads were replaced with the async coroutines. 

Please have another look when you find time.

marc-hb reviewed (commented):
Thanks for unifying the logic across Windows and Unix! This looks great, I think reducing that duplication will pay off from a maintenance perspective.

Wow, I didn't expect you to replace your Threads-based implementation with asyncio so quickly, I was just throwing an idea there. I admit I'm a bit disappointed by the complexity that seems required by `asyncio`; I naively expected the implementation would be somewhat simpler than with Threads... Are you sure you need a custom event loop? Can't "Streams" do the job without one? Also, are so many `try: expect OSerror` blocks required? Again, I didn't mind the threads; please correct me but performance shouldn't be an issue here so the less complex solution is probably the better one.

PS: your commit message is a bit out of date. No big deal.

MarkoSagadin commented:
So, the event loop is needed due to the `loop.sock_sendall` and `loop.sock_recv`, their api expects it. Keep in mind that `asyncio.get_running_loop()` just returns currently running loop, it doesn't create one.

I took a brief look at the `asyncio.open_connection` (Streams), but I missed the fact that I could also use those, I see now that I can provide an existing socket to the function.

The various `except OSError as e` are remnant from the Windows implementation: if rtt process exists due to some error or due to the Ctrl+C some graceful shut down is needed.

The performance is definitely not paramount here: both threads and asyncio seem to do just fine, there are no visible artifacts / glitches in the terminal.

You are right, the commit message needs an update, I have added few words about the current development.

---

Honestly, I am okay with either solution, I guess that both have some level of complexity which can be seen differently by people with different backgrounds. So please pick what you think would be lower maintenance burden in the long run. If needed I can still revert back to the thread implementation.

This PR was created some 4 months ago, the original code is from a PR that was opened on 30th of September. At this point I just want to get it merged.

marc-hb commented:
> Honestly, I am okay with either solution, I guess that both have some level of complexity which can be seen differently by people with different backgrounds. So please pick what you think would be lower maintenance burden in the long run. If needed I can still revert back to the thread implementation.

I think people (including myself) will be less familiar with the `async` model. It has big performance advantages in some contexts but this is not one of them. It can also require less code in some contexts but I'm surprised this does not seem to be one of them.

Looking back at your earlier Thread implementation as of https://github.com/zephyrproject-rtos/zephyr/commit/585e67b31cc5c443c741a2abccc11b0cf5d45af9, I just realized you probably don't require to start two new threads, starting a single Thread should be enough, no? So if you fork a (single) thread for `stdin`, then the main program can read on the socket and it can even use `select()` on both Windows and Unix! In case `select()` is "less blocking" than `recv()` and helps with KeyboardInterrupt and what not.

> This PR was created some 4 months ago, the original code is from a PR that was opened on 30th of September. At this point I just want to get it merged.

Sorry about that, as usual there is a maintainer shortage (here's a new one! #107507) and I'm not familiar with RTT. I haven't had the time to review line by line but to be honest I'm glad I didn't yet considering the high level design is still being decided and the code has been changing a lot. This being said, I never saw anything major and as soon as we get one or two "second opinions" to help settle on the design then things should accelerate.

EDIT: no promises but I suspect a single-Thread implementation could look the least review effort and help decide another reviewer to take a look.

BrandonDao commented:
Hi— I ran into the same bug with `rtt` independently ([issue #107637](https://github.com/zephyrproject-rtos/zephyr/issues/107637)) and had the same conclusion that the root cause of failure on Windows was the use of `select()` on a non-socket.

For what it's worth, I agree with @marc-hb that a simpler threading implementation (main thread + daemon for `stdin`) is likely better thanks to it being cleaner & more maintainable compared to the `asyncio` solution. It's actually the solution I arrived at when trying to roll my own workaround.
* I also agree with @marc-hb's point about using `select()` on the socket in the main thread.

Regardless of the concurrency model, the platform-appropriate stuff (`_windows_vt_console_mode` and `_unix_stdin_cbreak_mode`) feels self-contained + correct.

MarkoSagadin commented:
Hi @BrandonDao, thank you for interest in your PR. Yes, a while ago I have ran into the same issue as you did, but from a different angle: I wanted to get RTT over shell working :)

Hi @marc-hb! I have attempted to change the implementation to use a single thread + `select()` on the sock. While I got that working (RTT shell works fine), I noticed a strange side-effect: the Ctrl+C combination wasn't correctly working on the Windows, after I pressed it, nothing happened, only after I have pressed a random key the exception was raised. I have googled a bit, this seems to be connected with how `os.read()` behaves on Windows (it appears to ignore such key combination?!).

The two thread approach doesn't experience this issue. For some reason here the `stop.wait(timeout=0.1)` does the trick.

So I have added two additional commits to the PR (see the commit messages), that will also make it a bit easier for me to revert, once we decide on one:

1. First one uses single thread + `select()`
2. Second one uses two threads.

I have tested both implementations extensively, only the second one seems to be the "right" one, in the sense that it doesn't have any weird side effects.

(tbh, one more thread doesn't complicate the implementation that much, since the rtt loop is just moved to a different location, although I am not happy about those `not err.is_set() and not stop.is_set()` statements. They seem unnecessary, although they aren't.)

Please take a look when you find time.

---

> Sorry about that, as usual there is a maintainer shortage (here's a new one! https://github.com/zephyrproject-rtos/zephyr/pull/107507) and I'm not familiar with RTT. I haven't had the time to review line by line but to be honest I'm glad I didn't yet considering the high level design is still being decided and the code has been changing a lot. This being said, I never saw anything major and as soon as we get one or two "second opinions" to help settle on the design then things should accelerate.

No slight was meant against you, I am happy that I got a proper feedback from you :)

marc-hb commented:
> I have attempted to change the implementation to use a single thread + select() on the sock. While I got that working (RTT shell works fine), I noticed a strange side-effect: the Ctrl+C combination wasn't correctly working on the Windows, after I pressed it, nothing happened, only after I have pressed a random key the exception was raised. I have googled a bit, this seems to be connected with how os.read() behaves on Windows (it appears to ignore such key combination?!).

Interesting, it looks like SIGINT does not interrupt Windows system calls - unlike Unix system calls that get interrupted. AFAICT, SIGINT is not "ignored" when in Windows syscalls but you have to wait until the syscall completes. I found (too much) useful information on these pages: https://vorpus.org/blog/control-c-handling-in-python-and-trio/ , https://github.com/python-trio/trio/issues/42. I bet there are shorter pages but these ones seem extremely well informed (and too detailed).

So, instead of having the main thread doing nothing but waiting for the SIGINT, I think you could have one stdin thread + the main thread doing something like this:

```python

   while True:
      if select.select([sock_read], timeout=1-2 seconds,...)
         break
      # Ctrl-C / SIGINT gets received here with 1-2 seconds latency maximum
      # That comment is important!
```

It may look weird and counter-productive to manually "poll" select() in a loop, but I think it's actually a pretty good trade-off:
- non-zero but very low CPU usage when nothing happens: runs only once every 1-2 seconds and extremely briefly
- zero added network latency cause select() is still there doing its job
- Very decent Ctrl-C latency.
- Barely any additional code.

 
> (tbh, one more thread doesn't complicate the implementation that much, since the rtt loop is just moved to a different location, 

It's not just the complexity of the code, it's also the fact that the more threads there are and the more potential races and situations there are and the more painful it is to use a debugger or to debug generally speaking (another reason why I suggested async a while back but let's park that)

> although I am not happy about those not err.is_set() and not stop.is_set() statements. They seem unnecessary, although they aren't.)

I haven't looked at those in detail yet (sorry again), but I don't see anything shocking yet. Ideally, threads should always ask each other politely to stop. Interupting/killing at random points in time is always bad and the crazy length of https://vorpus.org/blog/control-c-handling-in-python-and-trio/ is a great demonstration of how bad random interruptions are :-)

20 years old throwback:
https://stackoverflow.com/questions/16504140/thread-stop-deprecated

MarkoSagadin commented:
> AFAICT, SIGINT is not "ignored" when in Windows syscalls but you have to wait until the syscall completes. 

Ah, the more you know. I have briefly looked at the blog post and the issue, but I will admit that I haven't read them completely. I do however understand the issues that can happen by terminating some process/thread unexpectedly. 


> So, instead of having the main thread doing nothing but waiting for the SIGINT, I think you could have one stdin thread + the main thread doing something like this:

This is actually the solution!

If I only add timeout parameter to the `sel.select()`, the Ctrl+C is correctly captured on Windows. I have did that in the [36a6df8](https://github.com/zephyrproject-rtos/zephyr/pull/100473/commits/36a6df8e8443c1e9d914006759cb503ece314b04). Only difference is that I decreased the timeout to 0.5, it feels a bit more snappier than 1 second.

I would be ok with going forward with the single thread plus select implementation. What do you think? I have rebased to the latest main and pushed mentioned timeout change.


> 20 years old throwback:
https://stackoverflow.com/questions/16504140/thread-stop-deprecated

I feel like that there should be (probably is, but I don't know it) some kind of "bible" book on the topic of threads and issues that can arise from them. I know of this: https://greenteapress.com/semaphores/LittleBookOfSemaphores.pdf

MarkoSagadin commented:
Naturally, a comment above `sel.select(timeout=0.5)`, that explains what is going on, should be added.

marc-hb commented on scripts/west_commands/runners/core.py:
```diff
@@ -1032,22 +1034,129 @@ def run_telnet_client(
+                stop.set()
+
+        stop = threading.Event()
+        err  = threading.Event()
```
Could you make these event names a tiny bit longer? Right now they are so short and generic that they look like things built in either Python or the threading module which makes the code a bit harder to read.

I mean things like `stop_requested` and `recv_error` - or anything better but just as long.

MarkoSagadin replied:
I know what you mean, they are a bit short.

I changed them to `stop_requested` and `error_received`.

MarkoSagadin commented:
@marc-hb 

I have added a comment and renamed the two variables.

marc-hb commented:
I just had a look at the 3 approaches and they all look OK! However, since you got the 1-thread approach working now, I think you can drop the 2 threads approach because why fork 2 threads when a single one does the job? Also, it's a bit surprising to have 3 different alternatives as commits in the same PR. Spoiled with choice! But that choice could make some reviewers wait until there are fewer alternatives left to review?

The commit with the 2-threads version https://github.com/zephyrproject-rtos/zephyr/commit/d43ea8b2574d7fe39f96faa696419233c2a42f23 will stay available for a long time (forever?) anyway even after you drop it from the PR.

I had another look at the `async` alternative  and it felt simpler than as last time... Did you simplify it a bit? Or maybe my brain got used to it somehow. I don't have a strong preference but I suspect most people would be more familiar with a thread. The async code is about the same size. I would love someone new to express a preference.

Impressive work anyway, thank you for all the efforts!

MarkoSagadin commented:
I went ahead and dropped the commits with async and two thread implementations, now the single thread + select remains.

> I had another look at the async alternative and it felt simpler than as last time... Did you simplify it a bit? Or maybe my brain got used to it somehow.

Nope, stayed the same :) 

> Impressive work anyway, thank you for all the efforts!

Thank you for kind words :)

I have also rebased against the latest main and force-pushed. 

>  I would love someone new to express a preference.

Maybe @mbolivar?

tpambor commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -418,13 +418,22 @@ def do_run(self, command, **kwargs):
                 self.run_telnet_client('localhost', self.rtt_port, sock,
                                        send_on_connect=rtt_config)
+            except ValueError:
+                # ValueError expection is thrown, if PC can't connect to the J-Link and
```
```suggestion
                # ValueError exception is thrown, if PC can't connect to the J-Link and
```
nit

tpambor commented on scripts/west_commands/runners/core.py:
```diff
@@ -1032,23 +1034,131 @@ def run_telnet_client(
+        ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004
+
+        stdin = kernel32.GetStdHandle(-10)
+        stdout = kernel32.GetStdHandle(-11)
```
```suggestion
        STD_INPUT_HANDLE = -10
        STD_OUTPUT_HANDLE = -11

        stdin = kernel32.GetStdHandle(STD_INPUT_HANDLE)
        stdout = kernel32.GetStdHandle(STD_OUTPUT_HANDLE)
```

tpambor reviewed (dismissed):
LGTM

MarkoSagadin commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -418,13 +418,22 @@ def do_run(self, command, **kwargs):
+                # we still try to open a socket to it.
+                # Silently capture the exception, the prompt from JLinkExe already
+                # contains good message about what happened.
+                pass
```
> and it's not clear what in the try-block would raise ValueError during a failed socket connect (socket.connect/send typically raise OSError).

Well, that is explained in the first sentence of my comment.

I replaced a `pass` with a debug log statement, so if any developer is debugging this part of the code here, the raised exception will be visible.

MarkoSagadin commented:
Thank you for approval @tpambor! I have addressed your comments and most of the Copilot's comments.

I didn't address two comments related to redirection of the stdin/stdout, since this isn't the usecase that is expected to be covered by the `west rtt` command.
